src/aind_metadata_upgrader/procedures_upgrade.py

```diff
@@ -0,0 +1,459 @@
+            "Fiber implant": self.subj_procedure_upgrader.upgrade_fiber_implant,
+            "Headframe": self.subj_procedure_upgrader.upgrade_headframe,
+            "Intra cerebellar ventricle injection":
+            self.subj_procedure_upgrader.upgrade_intra_cerebellar_ventricle_injection,
+            "Intra cisternal magna injection": self.subj_procedure_upgrader.upgrade_intra_cisternal_magna_injection,
+            "Intraperitoneal injection": self.subj_procedure_upgrader.upgrade_intraperitoneal_injection,
+            "Iontophoresis injection": self.subj_procedure_upgrader.upgrade_iontophoresis_injection,
+            "Nanoject injection": self.subj_procedure_upgrader.upgrade_nanoject_injection,
+            "Perfusion": self.subj_procedure_upgrader.upgrade_perfusion,
+            "Other subject procedure": self.subj_procedure_upgrader.upgrade_other_subject_procedure,
+            "Retro-orbital injection": self.subj_procedure_upgrader.upgrade_retro_orbital_injection,
+        }
+
+    def caller(self, func, model):
+        """Call a function with a model as an argument"""
+
+        return func(model)
+
+    def upgrade_subject_procedure(self, old_subj_procedure: dict):
+        """Map legacy SubjectProcedure model to current version"""
+
+        procedure_type = old_subj_procedure.get("procedure_type")
+        if procedure_type in self.upgrade_funcs.keys():
+
+            old_subj_procedure = drop_unused_fields(old_subj_procedure, procedure_type)
+
+            if old_subj_procedure.get("injection_materials"):
+                old_subj_procedure["injection_materials"] = InjectionMaterialsUpgrade(
+                    self.allow_validation_errors
+                ).upgrade_injection_materials(old_subj_procedure["injection_materials"])
+            else:
+                old_subj_procedure["injection_materials"] = [None]
+
+            return self.caller(self.upgrade_funcs[procedure_type], old_subj_procedure)
+        else:
+            logging.error(f"Procedure type {procedure_type} not found in list of procedure types")
+            return None
+
+    @staticmethod
+    def upgrade_specimen_procedure(self, old_specimen_procedure: Any) -> Optional[SpecimenProcedure]:
+        """Map legacy SpecimenProcedure model to current version"""
+
+        if type(old_specimen_procedure) is SpecimenProcedure:
+            return old_specimen_procedure
+        elif type(old_specimen_procedure) is dict and old_specimen_procedure.procedure_type is not None:
+            return SpecimenProcedure.model_validate(old_specimen_procedure)
+        else:
+            logging.error(f"Specimen procedure {old_specimen_procedure} passed in as invalid type")
+            return None
+
+    def upgrade_procedure(self) -> Optional[Procedures]:
+        """Map legacy Procedure model to current version"""
+
+        if semver.Version.parse(self._get_or_default(self.old_model, "schema_version", {})) <= semver.Version.parse(
```

Comment:
I think there is an `else` clause missing here.  What happens if you try to upgrade something that's already up to date?

Reply:
Right now, nothing. This class is only currently designed to upgrade procedures predating the `Surgery` update, so anything that we're passed in newer should be unchanged as of now.